openvdb_points/tools/IndexFilter.h

```diff
@@ -0,0 +1,495 @@
+    static LevelSetFilter create(const LeafT& leaf, const Data& data) {
+        return LevelSetFilter(data, AttributeHandle<openvdb::Vec3f>::create(leaf.constAttributeArray("P")));
+    }
+
+    template <typename IterT>
+    bool valid(const IterT& iter) const {
+        assert(mData);
+        assert(iter);
+
+        const openvdb::Coord ijk = iter.getCoord();
+        const openvdb::Vec3f voxelIndexSpace = ijk.asVec3d();
+
+        // Retrieve point position in voxel space
+        const openvdb::Vec3f& pointVoxelSpace = mPositionHandle->get(*iter);
+
+        // Compute point position in index space
+        const openvdb::Vec3f pointWorldSpace = mData->transform.indexToWorld(pointVoxelSpace + voxelIndexSpace);
+        const openvdb::Vec3f pointIndexSpace = mData->levelSetTransform.worldToIndex(pointWorldSpace);
+
+        // Perform level-set sampling
+        const typename LevelSetGridT::ValueType value = BoxSampler::sample(mData->accessor, pointIndexSpace);
+
+        // if min is greater than max, we invert so that values are valid outside of the range (not inside)
+        const bool invert = mData->min > mData->max;
+
+        return invert ? (value < mData->max || value > mData->min) : (value < mData->max && value > mData->min);
+    }
+
+private:
+    const Data* mData;
+    AttributeHandle<openvdb::Vec3f>::Ptr mPositionHandle;
+}; // class LevelSetFilter
+
+
+// BBox index filtering
+class BBoxFilter
+{
+public:
+    struct Data
+    {
+        Data(const openvdb::math::Transform& _transform,
+             const openvdb::BBoxd& _bboxWS)
+            : transform(_transform)
+            , bbox(transform.worldToIndex(_bboxWS)) { }
+        const openvdb::math::Transform& transform;
+        const openvdb::BBoxd bbox;
+    };
+
+    BBoxFilter( const Data& data,
+                const AttributeHandle<openvdb::Vec3f>::Ptr& positionHandle)
+        : mData(&data)
+        , mPositionHandle(positionHandle) { }
+
+    template <typename LeafT>
+    static BBoxFilter create(const LeafT& leaf, const Data& data) {
+        return BBoxFilter(data, AttributeHandle<openvdb::Vec3f>::create(leaf.constAttributeArray("P")));
+    }
+
+    template <typename IterT>
+    bool valid(const IterT& iter) const {
+        assert(mData);
+
+        const openvdb::Coord ijk = iter.getCoord();
+        const openvdb::Vec3f voxelIndexSpace = ijk.asVec3d();
+
+        // Retrieve point position in voxel space
+        const openvdb::Vec3f& pointVoxelSpace = mPositionHandle->get(*iter);
+
+        // Compute point position in index space
+        const openvdb::Vec3f pointIndexSpace = pointVoxelSpace + voxelIndexSpace;
+
+        return mData->bbox.isInside(pointIndexSpace);
+    }
+
+private:
+    const Data* mData;
+    const AttributeHandle<openvdb::Vec3f>::Ptr mPositionHandle;
+}; // class BBoxFilter
+
+
+// Index filtering based on evaluating both sub-filters
+template <typename T1, typename T2, bool And = true>
+class BinaryFilter
```

Comment:
Have you considered splitting this into an AndFilter and OrFilter to make its functionality a bit more explicit, especially in the default case where people will omit the third template argument? (at the cost of more code admittedly!)